lib/CPathfinder.cpp

```diff
@@ -412,17 +723,289 @@ bool CPathfinder::addTeleportOneWayRandom(const CGTeleport * obj) const
+}
+
+bool CPathfinderHelper::isLayerAvailable(const EPathfindingLayer layer) const
+{
+	switch(layer)
+	{
+	case EPathfindingLayer::AIR:
+		if(!options.useFlying)
+			return false;
+
+		break;
+
+	case EPathfindingLayer::WATER:
+		if(!options.useWaterWalking)
+			return false;
+
+		break;
+	}
+
+	return turnsInfo[turn]->isLayerAvailable(layer);
+}
+
+const TurnInfo * CPathfinderHelper::getTurnInfo() const
+{
+	return turnsInfo[turn];
+}
+
+bool CPathfinderHelper::hasBonusOfType(const Bonus::BonusType type, const int subtype) const
+{
+	return turnsInfo[turn]->hasBonusOfType(type, subtype);
+}
+
+int CPathfinderHelper::getMaxMovePoints(const EPathfindingLayer layer) const
+{
+	return turnsInfo[turn]->getMaxMovePoints(layer);
+}
+
+void CPathfinderHelper::getNeighbours(const CMap * map, const TerrainTile & srct, const int3 & tile, std::vector<int3> & vec, const boost::logic::tribool & onLand, const bool limitCoastSailing)
+{
+	static const int3 dirs[] = {
+		int3(-1, +1, +0),	int3(0, +1, +0),	int3(+1, +1, +0),
+		int3(-1, +0, +0),	/* source pos */	int3(+1, +0, +0),
+		int3(-1, -1, +0),	int3(0, -1, +0),	int3(+1, -1, +0)
+	};
+
+	for(auto & dir : dirs)
+	{
+		const int3 hlp = tile + dir;
+		if(!map->isInTheMap(hlp))
+			continue;
+
+		const TerrainTile & hlpt = map->getTile(hlp);
+		if(hlpt.terType == ETerrainType::ROCK)
+			continue;
+
+// 		//we cannot visit things from blocked tiles
+// 		if(srct.blocked && !srct.visitable && hlpt.visitable && srct.blockingObjects.front()->ID != HEROI_TYPE)
+// 		{
+// 			continue;
+// 		}
+
+		/// Following condition let us avoid diagonal movement over coast when sailing
+		if(srct.terType == ETerrainType::WATER && limitCoastSailing && hlpt.terType == ETerrainType::WATER && dir.x && dir.y) //diagonal move through water
+		{
+			int3 hlp1 = tile,
+				hlp2 = tile;
+			hlp1.x += dir.x;
+			hlp2.y += dir.y;
+
+			if(map->getTile(hlp1).terType != ETerrainType::WATER || map->getTile(hlp2).terType != ETerrainType::WATER)
+				continue;
+		}
+
+		if(indeterminate(onLand) || onLand == (hlpt.terType != ETerrainType::WATER))
+		{
+			vec.push_back(hlp);
+		}
+	}
+}
+
+int CPathfinderHelper::getMovementCost(const CGHeroInstance * h, const int3 & src, const int3 & dst, const TerrainTile * ct, const TerrainTile * dt, const int remainingMovePoints, const TurnInfo * ti, const bool checkLast)
+{
+	if(src == dst) //same tile
+		return 0;
+
+	if(!ti)
+		ti = new TurnInfo(h);
+
+	if(ct == nullptr || dt == nullptr)
+	{
+		ct = h->cb->getTile(src);
+		dt = h->cb->getTile(dst);
+	}
+
+	/// TODO: by the original game rules hero shouldn't be affected by terrain penalty while flying.
+	/// Also flying movement only has penalty when player moving over blocked tiles.
+	/// So if you only have base flying with 40% penalty you can still ignore terrain penalty while having zero flying penalty.
+	int ret = h->getTileCost(*dt, *ct, ti);
+	/// Unfortunately this can't be implemented yet as server don't know when player flying and when he's not.
```

Comment:
> What exactly is a problem?

Problem is fact that I have planned movement rework which is [explained on wiki](http://wiki.vcmi.eu/index.php?title=User:SXX/Pathfinding_and_movement). TL;DR: idea is to use pathfinder both for path calculation and validation. Validation won't require to create paths graph at all and going to be as fast as current server code which let us avoid code duplication.

In same code that branch already working even if costs are wrong. It's reasonable to merge working code and start new branch for other changes.